STM32/ACTenChannel/Core/Src/ACTenChannel.c

```diff
@@ -0,0 +1,395 @@
+
+/***************************************************************************************************
+** PRIVATE TYPEDEFS
+***************************************************************************************************/
+
+typedef struct ActuationInfo {
+    int pin; // pins 0-9 are interpreted as single ports - pin '-1' is interpreted as all
+    int pwmDutyCycle;
+    int timeOn; // time on is in seconds
+    bool isInputValid;
+} ActuationInfo;
+
+/***************************************************************************************************
+** PRIVATE FUNCTION DECLARATIONS
+***************************************************************************************************/
+
+static void CAallOn(bool isOn, int duration);
+static void CAportState(int port, bool state, int percent, int duration);
+static void userInput(const char *input);
+static void printAcTenChannelStatus();
+static void updateBoardStatus();
+
+/***************************************************************************************************
+** PRIVATE OBJECTS
+***************************************************************************************************/
+
+/* GPIO settings. */
+StmGpio heaterPorts[AC_TEN_CH_NUM_PORTS];
+static StmGpio fanCtrl;
+static StmGpio DQ1;
+static double heatSinkTemperature = 0; // Heat Sink temperature
+static bool isFanForceOn = false;
+
+static CAProtocolCtx caProto =
+{
+        .undefined = userInput,
+        .printHeader = CAPrintHeader,
+        .printStatus = printAcTenChannelStatus,
+        .jumpToBootLoader = HALJumpToBootloader,
+        .calibration = NULL,
+        .calibrationRW = NULL,
+        .logging = NULL,
+        .otpRead = CAotpRead,
+        .otpWrite = NULL,
+        .allOn = CAallOn,
+        .portState = CAportState,
+};
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Verbose print of the AC board status
+*/
+static void printAcTenChannelStatus()
+{
+    static char buf[600] = { 0 };
+    int len = 0;
+
+    len += snprintf(&buf[len], sizeof(buf) - len, "Fan     On: %d\r\n", stmGetGpio(fanCtrl));
+    for (int i = 0; i < AC_TEN_CH_NUM_PORTS; i++)
+    {
+        len += snprintf(&buf[len], sizeof(buf) - len, "Port %d: On: %d, PWM percent: %d\r\n", 
+                        i, stmGetGpio(heaterPorts[i]), getPWMPinPercent(i));
+    }
+
+    writeUSB(buf, len);
+}
+
+static void userInput(const char *input)
+{
+    if (strncmp(input, "fan on", 6) == 0)
+    {
+        isFanForceOn = true;
+        stmSetGpio(fanCtrl, true);
+    }
+    else if (strncmp(input, "fan off", 7) == 0)
+    {
+        isFanForceOn = false;
+        stmSetGpio(fanCtrl, false);
+    }
+    else 
+    {
+        HALundefined(input);
+    }
+}
+
+static void GpioInit()
+{
+    const int NO_PORTS = AC_TEN_CH_NUM_PORTS;
```

Comment:
Why do you redefine the number of ports and not use directly `AC_TEN_CH_NUM_PORTS` ?